### PR TITLE
check ALLOWED_DOMAINS with regex-match()

### PR DESCRIPTION
**Pull request by frankbits**
To be able to use this with preview deployment (different Domains for every branch), you would have to add them all to the Environment Variable. If that variable is checked with regex match(), it suffices to add the main part of the domains.

### CODE REVIEW

**Review by kyoshino — COMMENTED** [review]
Thanks for the PR! I previously thought I would add support for wildcard (`*`) characters, but regex matching is a better approach for sure 👍🏻 

We probably need to [escape the input](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions#escaping) because a dot in a domain name would be recognized as any single character in a regex. It may not be a practical issue but for example, if `ALLOWED_DOMAINS` is `www.example.com`, the authenticator would be free to use by `www-example.com`.

Other than that, it looks good!

File: src/index.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ export default {
       const domain = searchParams.get('site_id');
 
       // Check if the domain is whitelisted
-      if (ALLOWED_DOMAINS && !ALLOWED_DOMAINS.split(/,\s*/).includes(domain)) {
+      if (ALLOWED_DOMAINS && !ALLOWED_DOMAINS.split(/,\s*/).some(rx => domain.match(rx))) {
```
Comment (kyoshino) [inline src/index.js:11]:
```suggestion
      if (ALLOWED_DOMAINS && !ALLOWED_DOMAINS.split(/,\s*/).some(rx => domain.match(escapeRegExp(rx)))) {
```

File: src/index.js [hunk 2]
```diff
@@ -7,8 +7,10 @@ export default {
       const provider = searchParams.get('provider');
       const domain = searchParams.get('site_id');
 
+/** @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions#escaping */
+const escapeRegExp = (string) => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
```
Comment (kyoshino) [inline src/index.js:11]:
```suggestion
      /** @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions#escaping */
      const escapeRegExp = (string) => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

```

**Review by kyoshino — APPROVED** [review]

File: src/index.js [hunk 3]
```diff
@@ -7,8 +7,11 @@ export default {
       const provider = searchParams.get('provider');
       const domain = searchParams.get('site_id');
```
Comment (kyoshino) [inline src/index.js:9]:
```suggestion
```

File: src/index.js [hunk 4]
```diff
@@ -7,8 +7,11 @@ export default {
...
 
+      /** @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions#escaping */
+      const escapeRegExp = (string) => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+
       // Check if the domain is whitelisted
-      if (ALLOWED_DOMAINS && !ALLOWED_DOMAINS.split(/,\s*/).includes(domain)) {
+      if (ALLOWED_DOMAINS && !ALLOWED_DOMAINS.split(/,\s*/).some(rx => domain.match(escapeRegExp(rx)))) {
```
Comment (kyoshino) [inline src/index.js:14]:
```suggestion
      if (
        ALLOWED_DOMAINS &&
        !ALLOWED_DOMAINS.split(/,\s*/).some((rx) => domain.match(escapeRegExp(rx)))
      ) {
```

**Review by kyoshino — COMMENTED** [review]
Just formatting

**Comment by kyoshino** [comment]
Merging. Thanks again! 👍🏻

**Comment by kyoshino** [comment]
Hmm, I was probably wrong. It’s rather hard to use a regex like `[a-z]+.example.com` if the string is escaped. I’ll remove the escaping part in favour of your original code. 🙇🏻

**Comment by frankbits** [comment]
> Hmm, I was probably wrong. It’s rather hard to use a regex like `[a-z]+.example.com` if the strings are escaped. I’ll remove the escaping part in favour of your original code. 🙇🏻 

Didn't think about that, but yes. With the original Code you should still be able to use the `.` as a `.` by escaping it (`example\.com`).

Maybe it should normally not check the domain with regex, to prevent unwanted allowance, and only use regex-matching, if the domain is surrounded by regex-delimiters for example...

(`example.com,/[a-z]+\.example\.org/`)

also it would probably be better to use String.search(regex) instead of String.match(regex), as it works faster because it only searches for the first occurence (doesn't check rest of domain if `/[a-z]+\.example/` checks `forum.examplewebsite.org`)

```
!ALLOWED_DOMAINS.split(/,\s*/).some(rx => {
  if (/^\/.*\/$/.test(rx)) {
    return domain.match(escapeRegExp(rx)));
  }
  return domain == rx;
}
```
this implementation would not allow the use of regex-modifiers.

I don't think there would be cases were the domain has the `\` at the start and end of itself.  if there are, those would have to be entered with regex and escaped `\/`...

**Comment by frankbits** [comment]
> Hmm, I was probably wrong. It’s rather hard to use a regex like `[a-z]+.example.com` if the strings are escaped. I’ll remove the escaping part in favour of your original code. 🙇🏻 

Didn't think about that, but yes. With the original Code you should still be able to use the `.` as a `.` by escaping it (`example\.com`).

Maybe it should normally not check the domain with regex, to prevent unwanted allowance, and only use regex-matching, if the domain is surrounded by regex-delimiters for example...

(`example.com,/[a-z]+\.example\.org/`)

also it would probably be better to use String.search(regex) instead of String.match(regex), as it works faster because it only searches for the first occurence (doesn't check rest of domain if `/[a-z]+\.example/` checks `forum.examplewebsite.org`)

```
!ALLOWED_DOMAINS.split(/,\s*/).some(rx => {
  if (/^\/.*\/$/.test(rx)) {
    return domain.match(escapeRegExp(rx)));
  }
  return domain == rx;
}
```
this implementation would not allow the use of regex-modifiers.

I don't think there would be cases were the domain has the `\` at the start and end of itself.  if there are, those would have to be entered with regex and escaped `\/`...

Alternatively Regex-Matching could be toggled for all ALLIWED_DOMAINS by another Environment Variable (`DOMAIN_CHECK_REGEX` = `true`/`false`)

**Comment by frankbits** [comment]
If the code stays like it is now, the readme should have the `.` escaped (`example\.org`)

**Comment by frankbits** [comment]
> Hmm, I was probably wrong. It’s rather hard to use a regex like `[a-z]+.example.com` if the strings are escaped. I’ll remove the escaping part in favour of your original code. 🙇🏻 

Didn't think about that, but yes. With the original Code you should still be able to use the `.` as a `.` by escaping it (`example\.com`).

Maybe it should normally not check the domain with regex, to prevent unwanted allowance, and only use regex-matching, if the domain is surrounded by regex-delimiters for example...

(`example.com,/[a-z]+\.example\.org/`)

also it would probably be better to use String.search(regex) instead of String.match(regex), as it works faster because it only searches for the first occurence (doesn't check rest of domain if `/[a-z]+\.example/` checks `forum.examplewebsite.org`)

```
!ALLOWED_DOMAINS.split(/,\s*/).some(rx => {
  if (/^\/.*\/$/.test(rx)) {
    return domain.search(rx)
  }
  return domain == rx
}
```

Alternatively Regex-Matching could be toggled for all `ALLOWED_DOMAINS` by another Environment Variable (`DOMAIN_CHECK_REGEX` = `true`/`false`)

**Comment by frankbits** [comment]
Should the regex also only match, if the whole domain matches the regex?
Then the code would have to automatically add `^` and `$` if not already present...
(`/[a-z]+]\.example\.org/` -> /^[a-z]+]\.example\.org$/)
(`/^[a-z]+]\.example\.org$/` -> /^[a-z]+]\.example\.org$/)

Alternatively should the "String-matching" work more like the "regex-matching", by using `rx.includes(domain)` instead of `domain == rx`

Or those two just have a slightly different behaviour in this regard.

Either way the behaviour should be documented correctly in the README

**Comment by kyoshino** [comment]
Hey, I was thinking about this last night. The most common use case is matching all the subdomains, so maybe wildcards are enough and easy to write, just like [CSP source values](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources) allowing `*.example.com`. Internally the code will still use a regex to check the values. I’ll make the change now.

**Comment by kyoshino** [comment]
Done. README updated: https://github.com/sveltia/sveltia-cms-auth#step-3-configure-the-worker 🚀

**Pull request merged**